sonic_y_cable/y_cable_base.py

```diff
@@ -0,0 +1,1013 @@
+                One of the following predefined constants, the actual target to get the cursor values on:
+                     TARGET_NIC -> NIC,
+                     TARGET_TOR_A -> TORA,
+                     TARGET_TOR_B -> TORB
+        Returns:
+            a boolean, True if cursor values setting is successful
+                     , False if cursor values setting is not successful
+        """
+
+        raise NotImplementedError
+
+    def get_firmware_version(self, target):
+        """
+        This routine should return the active, inactive and next (committed)
+        firmware running on the target. Each of the version values in this context
+        could be a string with a major and minor number and a build value.
+        The port on which this API is called for can be referred using self.port.
+
+        Args:
+            target:
+                One of the following predefined constants, the actual target to get the firmware version on:
+                     TARGET_NIC -> NIC,
+                     TARGET_TOR_A -> TORA,
+                     TARGET_TOR_B -> TORB
+        Returns:
+            a Dictionary:
+                 with version_active, version_inactive and version_next keys
+                 and their corresponding values
+
+        """
+
+        raise NotImplementedError
+
+    def download_firmware(self, fwfile):
+        """
+        This routine should download and store the firmware on all the
+        components of the Y cable of the port for which this API is called..
+        This should include any internal transfers, checksum validation etc.
+        from TOR to TOR or TOR to NIC side of the firmware specified by the fwfile.
+        This basically means that the firmware which is being downloaded should be
+        available to be activated (start being utilized by the cable) once this API is
+        successfully executed.
+        Note that this API should ideally not require any rollback even if it fails
+        as this should not interfere with the existing cable functionality because
+        this has not been activated yet.
+        The port on which this API is called for can be referred using self.port.
+
+        Args:
+            fwfile:
+                 a string, a path to the file which contains the firmware image.
+                 Note that the firmware file can be in the format of the vendor's
+                 choosing (binary, archive, etc.). But note that it should be one file
+                 which contains firmware for all components of the Y-cable
+        Returns:
+            One of the following predefined constants:
+                FIRMWARE_DOWNLOAD_SUCCESS
+                FIRMWARE_DOWNLOAD_FAILURE
+
+                a predefined code stating whether the firmware download was successful
+                or an error code as to what was the cause of firmware download failure
+        """
+
+        raise NotImplementedError
+
+    def activate_firmware(self, fwfile=None):
```

Comment:
Is it necessary to add additional input parameter 'hitless' to enable/disable hitless upgrade 
```suggestion
    def activate_firrmware(self, fwfile=None, hitless=False):
```

- Hitless upgrade: it will backup/restore the current state (ex. variables of link status, API attributes...etc.) before and after firmware upgrade.

- Non-hitless upgrade: it will update the firmware regardless the current status, a link flip can be observed during the upgrade.

Ideally, the vendor should handle the hitless/non-hitless upgrade in their firmware or python API automatically, but this optional could be useful if we would like to do troubleshooting or reset the module reinitialize the chip especially for the y-cable which is not designed to issue a module reset (write lower page, byte93, bit7 to 1) from NIC end

Reply:
I will put in a param for hitless. 
Thanks

Reply:
fixed